### PR TITLE
Create separate lists of characters that are allowed within a tenant id vs scopes in AzureCliCredential.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Follow-up to: https://github.com/Azure/azure-sdk-for-cpp/pull/5085

### CODE REVIEW

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,8 @@
 
 ### Other Changes
 
+- Create separate lists of characters that are allowed within a tenant id vs scopes in `AzureCliCredential`.
```
Comment (antkmsft) [inline sdk/identity/azure-identity/CHANGELOG.md:13]: I'd rephrase to remove `vs`

File: sdk/identity/azure-identity/src/azure_cli_credential.cpp [hunk 2]
```diff
@@ -120,8 +113,12 @@ AzureCliCredential::AzureCliCredential(const Core::Credentials::TokenCredentialO
...
     const
 {
-  ThrowIfNotSafeCmdLineInput(scopes, "Scopes");
-  ThrowIfNotSafeCmdLineInput(m_tenantId, "TenantID");
+  // The OAuth 2.0 RFC (https://datatracker.ietf.org/doc/html/rfc6749#section-3.3) allows space as
+  // well for a list of scopes, but that isn't currently required.
+  ThrowIfNotSafeCmdLineInput(
+      scopes, ".-:/_", "Scopes"); // Characters allowed in the scope are [0-9a-zA-Z-.:/_].
+  ThrowIfNotSafeCmdLineInput(
+      m_tenantId, ".-", "TenantID"); // Characters allowed in the tenant id are [0-9a-zA-Z-.].
```
Comment (antkmsft) [inline sdk/identity/azure-identity/src/azure_cli_credential.cpp:121]: nit: I find these "Characters allowed in the ... are ..." a to excessive - you write in the comment what the code clearly does in the very same line - not it is only obvious, but also the comment may get out of sync and become confusing.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 3]
```diff
@@ -344,6 +344,50 @@ TEST(AzureCliCredential, UnsafeChars)
...
+    trc.Scopes.push_back(std::string("https://storage.azure.com/.default"));
+    EXPECT_THROW(static_cast<void>(azCliCred.GetToken(trc, {})), AuthenticationException);
+
+    try
+    {
+      auto const token = azCliCred.GetToken(trc, {});
+    }
+    catch (AuthenticationException const& e)
+    {
+      EXPECT_TRUE(std::string(e.what()).find("Unsafe") != std::string::npos) << e.what();
```
Comment (antkmsft) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:367]: I'd go and hardcode the expected exception message, if not the whole message - then the larger part. After all, error messages, and diagnosability is also a feature we deliver, therefore it should be tested, and it is totally normal to expect a certain text here. It is the Identity library code that makes the error message, so it is under our control. If we change the error message, we update the test.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 4]
```diff
@@ -344,6 +344,50 @@ TEST(AzureCliCredential, UnsafeChars)
   }
 }
 
+TEST(AzureCliCredential, UnsafeCharsScopeAndTenantId)
+{
+  std::string const ValidScopeButNotTenantId = "string:withcolon";
```
Comment (antkmsft) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:349]: I'd verify the whole set of 5 characters - first two accepted at both, last three are only allowed in scope. It can be a single unit test (no need for separate testcases), but I would use copy&paste to repeat some parts of the same test below, and not use a loop so that the failure message is clearer.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp [hunk 5]
```diff
@@ -344,6 +344,50 @@ TEST(AzureCliCredential, UnsafeChars)
...
+    trc.Scopes.push_back(
+        std::string("https://storage.azure.com/.default") + ValidScopeButNotTenantId);
+
+    try
+    {
+      auto const token = azCliCred.GetToken(trc, {});
+    }
+    catch (AuthenticationException const& e)
+    {
+      EXPECT_TRUE(std::string(e.what()).find("Unsafe") == std::string::npos) << e.what();
```
Comment (antkmsft) [inline sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:386]: I'd at least find `"Unsafe command line input found in Scopes."`, if not the whole message including the credential prefix.

**Review by antkmsft — APPROVED** [review]

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp (hunk 3 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:367]: We should come up with guidelines on this. Testing for exact error messages adds brittleness to unit testing (unnecessary tight coupling), and historically (from my .NET experience), we'd avoid doing such things.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp (hunk 5 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:386]: The only reason I am doing so here, even partially, is to explicitly check that the exception is happening because of correct validation of unsupported characters and not because the artifact that we'd fail with AuthenticationException in CI for other reasons (even for valid tenant id/scopes). Otherwise, the test was passing even if we don't do any string validation.

My intention is not to verify exact string content of the message (which is an open topic to close on separately).

Ideally, I would prefer not to assert for any string literal.

File: sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp (hunk 4 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_cli_credential_test.cpp:349]: I can fairly easily test for all the not allowed characters.

Validating allowed characters would still fail in CI for unrelated reasons, and I am going to leave out unit tests enhancements/refactoring as a separate work item, to avoid scope creep here.
